othello/apps/games/views.py

```diff
@@ -0,0 +1,127 @@
+import json
+import logging
+
+from django.conf import settings
+from django.contrib import messages
+from django.contrib.auth.decorators import login_required
+from django.http import FileResponse
+from django.shortcuts import get_object_or_404, redirect, render
+from django.utils import timezone
+
+from .forms import DownloadSubmissionForm, GameForm, SubmissionForm
+from .models import Game
+from .utils import serialize_game_info
+
+logger = logging.getLogger("othello")
+
+
+@login_required
+def upload(request):
+    if request.method == "GET":
+        return render(
+            request,
+            "games/upload.html",
+            {
+                "success": False,
+                "submission_form": SubmissionForm(),
+                "change_form": DownloadSubmissionForm(request.user),
+            },
+        )
+    action, success = request.POST.get("action", False), False
+    if action == "new_submission":
+        form = SubmissionForm(request.POST, request.FILES)
+        if form.is_valid():
+            try:
+                submission = form.save(commit=False)
+                submission.user = request.user
+                submission.save()
+                success = True
+            except BaseException as e:
+                messages.error(
+                    request,
+                    "Unable to upload script at this time, try again later",
+                    extra_tags="danger",
+                )
+                raise e
+        else:
+            for errors in form.errors.get_json_data().values():
+                for error in errors:
+                    messages.error(request, error["message"], extra_tags="danger")
+    elif action == "download_submission":
```

Comment:
This should be split into a separate `download` page. Then you can avoid the `action` mess and just run the upload code if `request.method == POST`.

Reply:
wdym separate `download` page? Do you mean there should be a separate view for downloading submission. I like the upload and download forms being on the same page

Reply:
The forms can be on the same page, but for the actual downloading there should be a separate page. In fact, it should use a `GET` request.

Reply:
Ah, so make another url endpoint for downloads?

In that case, can't I just change the action url of the download form to the download url? Submitting the form would still be a `POST` request b/c the user would be submitting a form. I would like to keep using the download form b/c it takes care of security (Can't download another users' code b/c other users' Submission ids aren't a choice in the form field)

Reply:
Um, no, this shouldn't be a POST request. POST requests are 1) for actions that shouldn't be repeated carelessly (like creating a DB entry) or 2) requests where the data passed needs to be protected. This fits neither.

Remove the `require_POST` decorator; change the `method` to `GET` and remove the `{% csrf_token %}` in the template.

Reply:
How will that work with `DownloadForm`?

Reply:
?

Reply:
Submitting `DownloadForm` is a `POST` request

Reply:
Just switch it to a `GET` request by following the steps I described above. 99% of the time, downloading information should be a `GET` request.